### PR TITLE
Fix transpose and sum batching rule

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47632 Batching rule for torch.squeeze(tensor)
* **#47628 Fix transpose and sum batching rule**
* #47625 Make kwargs argument optional in _batched_grad_test
* #47621 Tensor.contiguous, Tensor.is_contiguous batch rule

Pytorch has a special case where scalar_tensor.transpose(dim0, dim1) works
if dim0, dim1 are either 0 or -1 and returns the scalar tensor. If the following happens:
```py
>>> x = torch.randn(B0)  # the per-examples are all scalars
>>> vmap(lambda x: x.transpose(0, 0), x)
>>> vmap(lambda x: x.transpose(0, -1), x)
>>> vmap(lambda x: x.transpose(-1, 0), x)
>>> vmap(lambda x: x.transpose(-1, -1), x)
```
then we replicate this behavior.

This PR also fixes the behavior for the sum batching rule. For the sum
batching rule, I had forgotten to take into account the -1 case:
```py
>>> x = torch.randn(B0)  # the per-examples are all scalars
>>> vmap(lambda x: x.sum(0), x)  # OK
>>> vmap(lambda x: x.sum(-1), x) # fails before this PR

```

Test Plan:
- new tests

Differential Revision: [D24843658](https://our.internmc.facebook.com/intern/diff/D24843658)

### CODE REVIEW

**Review by albanD — DISMISSED** [review]
lgtm

**Review by albanD — APPROVED** [review]
So many edge cases...
Should the dims be wrapped at a higher level to avoid the issue with the -1 case?

**Comment by zou3519** [comment]
> So many edge cases...
> Should the dims be wrapped at a higher level to avoid the issue with the -1 case?

Wrapping the dims in `getPhysicalDim` would work for transpose, but for reduction functions like `sum` it's a little more tricky. There's no way to tell `torch.sum(x, dims)` to "don't do any reduction over x and instead make a clone of x". In numpy we could have done `torch.sum(x, dims=[])`, but that doesn't work in PyTorch: #29137

**Comment by albanD** [comment]
> There's no way to tell torch.sum(x, dims) to "don't do any reduction over x and instead make a clone of x".

Not sure to follow here. In which that wouldn't work? If it has a 0 dim dimension, I guess the regular sum will do the right thing? That feels unrelated to the dim wrapping?

**Comment by zou3519** [comment]
> > There's no way to tell torch.sum(x, dims) to "don't do any reduction over x and instead make a clone of x".
> 
> Not sure to follow here. In which that wouldn't work? If it has a 0 dim dimension, I guess the regular sum will do the right thing? That feels unrelated to the dim wrapping?

The hypothetical I was going through is that if in the batching rule, we only did the following:
- perform dim wrapping
- then call `at::sum` in a way that would give the correct semantics. Something like:
```cpp
Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, ...) {
  auto self_physical = MultiBatchVmapTransform::logicalToPhysical(self);
  auto dims_physical = // ... get the physical dims and do some dim wrapping 
  auto result = at::sum(self_physical.tensor(), dims_physical, keepdim, dtype);
  return self_physical.newLogicalFromPhysical(result);
}
```
In this case, if `self` is a batch of scalars, then (assuming a single batch dim) the physical tensor has shape `[B0]`. The correct semantics here is to not perform any reduction. This would be possible if we say that `physical_dims = []` (so, we're telling torch.sum to not do any reduction). However, the semantics of `torch.sum(x, dims=[])` is actually to perform a full reduce instead of no reduction.

**Comment by albanD** [comment]
I see ok!
Sounds good!
